src/ray/core_worker/task_event_buffer.h

```diff
@@ -66,10 +70,99 @@ class TaskEventBuffer {
   ///       CoreWorker disconnects to ensure all task events in the buffer are sent.
   virtual void FlushEvents(bool forced) = 0;
 
+  /// Start the TaskEventBuffer.
+  ///
+  /// Connects the GCS client, starts its io_thread, and sets up periodical runner for
+  /// flushing events to GCS.
+  /// \param auto_flush Test only flag to disable periodical flushing events if false.
+  /// \return If setup succeeds. When failure, events will not be reported.
+  virtual Status Start(bool auto_flush = true) = 0;
+
   /// Stop the TaskEventBuffer and it's underlying IO, disconnecting GCS clients.
   virtual void Stop() = 0;
 };
 
+/// Implementation of TaskEventBuffer.
+///
+/// The buffer has its own io_context and io_thread, that's isolated from other
+/// components.
+///
+/// This class is thread-safe.
+class TaskEventBufferImpl : public TaskEventBuffer {
+ public:
+  /// Constructor
+  ///
+  /// \param gcs_client GCS client
+  TaskEventBufferImpl(std::unique_ptr<gcs::GcsClient> gcs_client);
+
+  void AddTaskEvents(rpc::TaskEvents task_events) LOCKS_EXCLUDED(mutex_) override;
+
+  void FlushEvents(bool forced) LOCKS_EXCLUDED(mutex_) override;
+
+  Status Start(bool auto_flush = true) LOCKS_EXCLUDED(mutex_) override;
+
+  /// Stop the TaskEventBuffer and it's underlying IO, disconnecting GCS clients.
+  void Stop() LOCKS_EXCLUDED(mutex_) override;
+
+ private:
+  std::vector<rpc::TaskEvents> GetAllTaskEvents() LOCKS_EXCLUDED(mutex_) {
+    absl::MutexLock lock(&mutex_);
+    std::vector<rpc::TaskEvents> copy(buffer_);
+    return copy;
+  }
+
+  size_t GetNumTaskEventsDropped() LOCKS_EXCLUDED(mutex_) {
+    absl::MutexLock lock(&mutex_);
+    return num_task_events_dropped_;
+  }
+
+  /// Mutex guarding task_events_data_.
+  absl::Mutex mutex_;
+
+  /// IO service event loop owned by TaskEventBuffer.
+  instrumented_io_context io_service_;
+
+  /// Work guard to prevent the io_context from exiting when no work.
+  boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard_;
+
+  /// Dedicated io thread for running the periodical runner and the GCS client.
+  std::thread io_thread_;
+
+  /// The runner to run function periodically.
+  PeriodicalRunner periodical_runner_;
+
+  /// Client to the GCS used to push profile events to it.
+  std::unique_ptr<gcs::GcsClient> gcs_client_ GUARDED_BY(mutex_);
+
+  /// Buffered task events.
+  std::vector<rpc::TaskEvents> buffer_ GUARDED_BY(mutex_);
+
+  /// A iterator into buffer_ that determines which element to be overwritten.
+  size_t next_idx_to_overwrite_ GUARDED_BY(mutex_) = 0;
+
+  size_t num_task_events_dropped_ GUARDED_BY(mutex_) = 0;
+
+  /// Thread local fields accessed only in io_thread_.
+
+  /// True if there's a pending gRPC call. It's a simple way to prevent overloading
+  /// GCS with too many calls. There is no point sending more events if GCS could not
+  /// process them quick enough.
+  bool grpc_in_progress_ = false;
```

Comment:
I'm not sure I believe this is only accessed by the IO Thread--- if this is the case, could we add an assert that the thread id of FlushEvents() always equals the io thread one then?

Reply:
Ah you are right - while periodic invocations to `FlushEvents` only in io_thread, we do call `FlushEvents` in main thread when disconnecting.